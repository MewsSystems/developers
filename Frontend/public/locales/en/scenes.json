{
  "home": {
    "title": "Home",
    "description": "Amazing Movie DB app"
  }
}
