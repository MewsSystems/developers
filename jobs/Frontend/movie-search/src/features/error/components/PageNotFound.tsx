import { Stack, Typography } from '@mui/material';

export default function PageNotFound() {
  return (
    <Stack data-testid="page-not-found-content" direction="column" alignItems="center" spacing={4}>
      <Typography variant="h1">{'(╯°□°)╯︵ ┻━┻'}</Typography>
      <Typography sx={{ fontWeight: 'bold' }} variant="h2" color="error.main">
        404
      </Typography>
      <Typography variant="h3">Something went wrong! This page could not be found.</Typography>
    </Stack>
  );
}
