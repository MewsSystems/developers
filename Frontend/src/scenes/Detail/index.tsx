import { SceneMeta } from 'components/SceneMeta'
import { RouteComponentProps } from '@reach/router'
import { useGetMovieDetail } from 'hooks/useGetMovieDetail'
import React from 'react'
import { HeroBanner } from 'components/HeroBanner'
import { Heading } from 'components/Heading'
import { COLORS } from 'constants/colors'
import styled from 'styled-components'
import { Credits } from './components/Credits'
import { Content } from 'components/Content'
import { ErrorContent } from 'components/ErrorContent'
import { useTranslation } from 'react-i18next'
import { Rating } from 'components/Rating'
import { Loading } from './components/Loading'
import { Info } from './components/Info'

const StyledHeroBanner = styled(HeroBanner)`
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  padding: 2rem;
`

const Title = styled(Heading)`
  margin-bottom: 0.2rem;
  text-shadow: 1px 1px 0 ${COLORS.BLACK};
`

const Overview = styled.p`
  max-width: 45rem;
  margin: 1rem 0 0;
  color: ${COLORS.WHITE};
  text-shadow: 1px 1px 0 ${COLORS.BLACK};
`

const Grid = styled.div`
  display: grid;
  grid-gap: 2rem;
  grid-template-columns: auto 24rem;
`

export interface DetailProps extends RouteComponentProps {
  id?: number
}

export const Detail: React.FC<DetailProps> = ({ id }) => {
  const { t } = useTranslation()
  const { movieDetail, isLoading, hasErrored } = useGetMovieDetail(id)

  if (hasErrored)
    return (
      <ErrorContent
        title={t('error-content.title')}
        text={t('error-content.text')}
      />
    )
  if (isLoading) return <Loading />

  if (movieDetail) {
    const {
      backdrop_path,
      title,
      overview,
      vote_average,
      vote_count,
    } = movieDetail

    return (
      <>
        <SceneMeta title={title} description={overview} />
        <StyledHeroBanner background={backdrop_path && backdrop_path}>
          <Content>
            <Title level={1} color={COLORS.WHITE}>
              {title}
            </Title>
            <Rating rating={vote_average / 2} total={vote_count} />
            <Overview>{overview}</Overview>
          </Content>
        </StyledHeroBanner>
        <Content>
          <Grid>
            <Credits id={id} />
            <Info data={movieDetail} />
          </Grid>
        </Content>
      </>
    )
  }

  return null
}
