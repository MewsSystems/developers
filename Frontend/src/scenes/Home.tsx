import React, { useEffect, useState, useCallback } from 'react'
import { RouteComponentProps, navigate } from '@reach/router'
import { SceneMetaProps, SceneMeta } from 'components/SceneMeta'
import { Heading } from 'components/Heading'
import { useTranslation } from 'react-i18next'
import { useThunkDispatch } from 'hooks/useThunkDispatch'
import { hydrateTrendingMovies } from 'state/actions/movies'
import { useSelector } from 'react-redux'
import { State } from 'state/rootReducer'
import { differenceInDays } from 'date-fns'
import styled from 'styled-components'
import { MovieCard } from 'components/MovieCard'
import { Movie } from 'model/api/Movie'
import { Button } from 'components/Button'
import { useGetImagePath } from 'hooks/useGetImagePath'
import { ConfigurationBackdropSizesEnum } from 'model/api/ConfigurationBackdropSizesEnum'
import { Card as LoadingCard } from 'components/Loading/Card'
import { Spacing } from 'components/Spacing'
import { Content } from 'components/Content'

const Grid = styled.div`
  display: grid;
  grid-gap: 1rem;
  grid-template-columns: 1fr 1fr 1fr;
`

const { PUBLIC_URL } = process.env

export const Home: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => {
  const { t } = useTranslation('home')
  const dispatch = useThunkDispatch()
  const [trendingMovies, setTrendingMovies] = useState<Movie[]>([])
  const [page, setPage] = useState(1)
  const {
    updatedAt,
    trending: { results, total_pages },
    loading,
  } = useSelector((state: State) => state.movies)
  const getImagePath = useGetImagePath(ConfigurationBackdropSizesEnum.W_780)
  const dayDifference = differenceInDays(new Date(), updatedAt)

  const handleLoadMore = useCallback(() => {
    setPage(page + 1)
  }, [page])

  const reduceTrendingMovies = useCallback(() => {
    return Object.keys(results).reduce<Movie[]>((acc, index) => {
      if (Number(index) <= page) {
        acc = [...acc, ...results[index]]
      }

      return acc
    }, [])
  }, [results, page])

  const handleCardClick = useCallback((id: number) => {
    navigate(`${PUBLIC_URL}/movie/${id}`)
  }, [])

  useEffect(() => {
    if (dayDifference > 7) {
      dispatch(hydrateTrendingMovies(true))
    } else {
      dispatch(hydrateTrendingMovies(false, page))
    }
  }, [dispatch, dayDifference, page])

  useEffect(() => {
    setTrendingMovies(reduceTrendingMovies())
  }, [page, results, reduceTrendingMovies])

  return (
    <>
      <SceneMeta title={title} description={description} />
      <Content>
        <Heading level={1}>{t('trending')}</Heading>
        {trendingMovies.length > 0 && (
          <>
            <Grid>
              {trendingMovies.map(
                ({ title, overview, backdrop_path, original_language, id }) => (
                  <MovieCard
                    key={id}
                    id={id}
                    title={title}
                    overview={overview}
                    background={getImagePath(backdrop_path)}
                    language={original_language}
                    onClick={handleCardClick}
                  />
                )
              )}
              {loading && (
                <>
                  <LoadingCard />
                  <LoadingCard />
                  <LoadingCard />
                  <LoadingCard />
                  <LoadingCard />
                </>
              )}
            </Grid>
            <Spacing outer="2rem 0 0">
              <Button
                onClick={handleLoadMore}
                disabled={page === total_pages || loading}
                fullWidth
              >
                {t('load-more')}
              </Button>
            </Spacing>
          </>
        )}
      </Content>
    </>
  )
}
