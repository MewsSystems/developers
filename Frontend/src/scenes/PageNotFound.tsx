import React from 'react'
import { RouteComponentProps } from '@reach/router'
import { SceneMetaProps, SceneMeta } from 'components/SceneMeta'
import { Heading } from 'components/Heading'
import { Content } from 'components/Content'

export const PageNotFound: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => (
  <>
    <SceneMeta title={title} description={description} />
    <Content>
      <Heading level={1}>Page not found</Heading>
    </Content>
  </>
)
