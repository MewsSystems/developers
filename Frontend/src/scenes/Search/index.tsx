import React, { useState, useCallback, useEffect } from 'react'
import { SceneMeta, SceneMetaProps } from 'components/SceneMeta'
import { State } from 'state/rootReducer'
import { useSelector } from 'react-redux'
import { RouteComponentProps, navigate } from '@reach/router'
import { Content } from 'components/Content'
import { Heading } from 'components/Heading'
import { useTranslation } from 'react-i18next'
import { COLORS } from 'constants/colors'
import styled from 'styled-components'
import { MovieList } from 'components/MovieList'
import { Movie } from 'model/api/Movie'
import { useGetSearchResults } from 'hooks/useGetSearchResults'
import { Spacing } from 'components/Spacing'
import { ErrorContent } from 'components/ErrorContent'
import { Loading } from './components/Loading'

const { PUBLIC_URL } = process.env

const SearchQuery = styled.span`
  color: ${COLORS.DARK_GRAY};
  font-weight: bold;
`
export const Search: React.FC<RouteComponentProps<SceneMetaProps>> = ({
  title,
  description,
}) => {
  const { t } = useTranslation('search')
  const [movies, setMovies] = useState<Movie[]>([])
  const [page, setPage] = useState(1)
  const [totalPages, setTotalPages] = useState(0)
  const searchValue = useSelector((state: State) => state.search.value)
  const { results, isLoading, hasErrored } = useGetSearchResults(page)

  const handleLoadMore = useCallback(() => {
    setPage(page + 1)
  }, [page])

  const handleMovieClick = useCallback((id: number) => {
    navigate(`${PUBLIC_URL}/movie/${id}`)
  }, [])

  useEffect(() => {
    setMovies(state => [...state, ...results.results])
    setTotalPages(results.total_pages)
  }, [results])

  if (hasErrored)
    return (
      <ErrorContent
        title={t('error-content.title')}
        text={t('error-content.text')}
      />
    )

  if (isLoading) return <Loading />

  return (
    <>
      <SceneMeta title={`${title} ${searchValue}`} description={description} />
      <Content>
        <Heading level={1}>{t('search-results')}</Heading>
        {searchValue.length > 0 ? (
          <>
            <Spacing outer="0 0 2rem">
              {t('searching-for')} <SearchQuery>{searchValue}</SearchQuery>
            </Spacing>

            {results.total_results === 0 ? (
              <Heading level={2}>
                {t('no-search-results', { searchValue })}
              </Heading>
            ) : (
              <MovieList
                items={movies}
                onMovieClick={handleMovieClick}
                onLoadMore={handleLoadMore}
                page={page}
                totalPages={totalPages}
                loading={isLoading}
              />
            )}
          </>
        ) : (
          <Heading level={2}>{t('no-search-value')}</Heading>
        )}
      </Content>
    </>
  )
}
