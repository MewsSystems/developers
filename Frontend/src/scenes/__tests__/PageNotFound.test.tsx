import React from 'react'
import { shallow } from 'enzyme'
import { PageNotFound } from 'scenes/PageNotFound'

describe('Test PageNotFound scene', () => {
  it('Tests snapshot', () => {
    expect(shallow(<PageNotFound />)).toMatchSnapshot()
  })
})
