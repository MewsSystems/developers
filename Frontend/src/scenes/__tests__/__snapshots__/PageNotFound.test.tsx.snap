// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test PageNotFound scene Tests snapshot 1`] = `
<Fragment>
  <SceneMeta />
  <Heading
    level={1}
  >
    Page not found
  </Heading>
</Fragment>
`;
